src/core/json/path.cc

```diff
@@ -0,0 +1,234 @@
+  };
+
+  unsigned matches_ = 0;
+};
+
+auto Dfs::Item::Advance(const PathSegment& segment) -> AdvanceResult {
+  AdvanceResult result = std::visit(  // line break
+      Overload{
+          [&](monostate) { return Init(segment); },  // Init state
+          [&](JsonType::const_object_iterator& it) -> AdvanceResult {
+            if (!ShouldIterateAll(segment.type()))
+              return nullptr;  // exhausted
+
+            ++it;
+            if (it == obj_->object_range().end()) {
+              return nullptr;  // exhausted
+            }
+            return &it->value();
+          },
+          [&](JsonType::const_array_iterator& it) -> AdvanceResult {
+            if (!ShouldIterateAll(segment.type()))
+              return nullptr;
+
+            ++it;
+            if (it == obj_->array_range().end()) {
+              return nullptr;
+            }
+            return &*it;
+          },
+      },
+      state_);
+  return result;
+}
+
+auto Dfs::Item::Init(const PathSegment& segment) -> AdvanceResult {
+  switch (segment.type()) {
+    case SegmentType::IDENTIFIER: {
+      if (obj_->is_object()) {
+        auto it = obj_->find(segment.identifier());
+        if (it != obj_->object_range().end()) {
+          state_ = it;
+          return &it->value();
+        } else {
+          return nullptr;  // exhausted
+        }
+      }
+      break;
+    }
+    case SegmentType::INDEX: {
+      unsigned index = segment.index();
+      if (obj_->is_array()) {
+        if (index >= obj_->size()) {
+          return make_unexpected(OUT_OF_BOUNDS);
+        }
+        auto it = obj_->array_range().cbegin() + index;
+        state_ = it;
+        return &*it;
+      }
+      break;
+    }
+
+    case SegmentType::WILDCARD: {
+      if (obj_->is_object()) {
+        jsoncons::range rng = obj_->object_range();
+        if (rng.cbegin() == rng.cend()) {
+          return nullptr;
+        }
+        state_ = rng.begin();
+        return &rng.begin()->value();
+      }
+
+      if (obj_->is_array()) {
+        jsoncons::range rng = obj_->array_range();
+        if (rng.cbegin() == rng.cend()) {
+          return nullptr;
+        }
+        state_ = rng.cbegin();
+        return &*rng.cbegin();
+      }
+      break;
+    }
+  }  // end switch
+
+  return make_unexpected(MISMATCH);
+}
+
+void Dfs::Traverse(absl::Span<const PathSegment> path, const JsonType& root, const Cb& callback) {
+  DCHECK(!path.empty());
+  if (path.size() == 1) {
+    PerformStep(path[0], root, callback);
+    return;
+  }
+
+  vector<Item> stack;
+  stack.emplace_back(&root);
+
+  do {
+    unsigned depth = stack.size();
+    // init or advance the current object
+    AdvanceResult res = stack.back().Advance(path[depth - 1]);
+    if (res && *res != nullptr) {
+      const JsonType* next = *res;
+      DVLOG(2) << "Handling now " << next->to_string();
+
+      if (depth + 1 < path.size()) {
```

Comment:
It will likely break with recursive descent, otherwise you need to store a multi-depth iterator transparently, at which point you already have two systems doing the same

Reply:
Yes, it will break it. I will fix it in the subsequent PRs. I thought to move depth to be part of `Item` object. 
Please tell me if you have better ideas.